state/futures/selectors.ts

```diff
@@ -136,26 +136,72 @@ export const selectFuturesAccount = createSelector(
 );
 
 export const selectCrossMarginPositions = createSelector(
+	selectPrices,
 	(state: RootState) => state.futures,
-	(futures) => {
+	(prices, futures) => {
 		return futures.crossMargin.account && futures.crossMargin.positions[futures.crossMargin.account]
 			? futures.crossMargin.positions[futures.crossMargin.account].map(
 					// TODO: Maybe change to explicit serializing functions to avoid casting
-					(p) => deserializeWeiObject(p, futuresPositionKeys) as FuturesPosition
+					(p) => {
+						const positionDetails = deserializeWeiObject(p, futuresPositionKeys) as FuturesPosition;
+
+						const position = positionDetails.position;
+						const price = prices[positionDetails.asset]?.offChain ?? position?.lastPrice;
+
+						const pnl = position?.size.mul(
+							position.lastPrice.sub(price).mul(position.side === PositionSide.LONG ? -1 : 1)
+						);
+						const pnlPct = pnl?.div(position?.initialMargin);
+
+						// update unrealized pnl with offchain price
+						return {
+							...positionDetails,
+							position: position
+								? {
+										...position,
+										pnl: pnl ?? position.pnl,
+										pnlPct: pnlPct ?? position.pnlPct,
+								  }
+								: null,
+						};
+					}
 			  )
 			: [];
 	}
 );
 
 export const selectIsolatedMarginPositions = createSelector(
 	selectWallet,
+	selectPrices,
 	(state: RootState) => state.futures,
-	(wallet, futures) => {
+	(wallet, prices, futures) => {
 		if (!wallet) return [];
 		return futures.isolatedMargin.positions[wallet]
 			? futures.isolatedMargin.positions[wallet].map(
 					// TODO: Maybe change to explicit serializing functions to avoid casting
-					(p) => deserializeWeiObject(p, futuresPositionKeys) as FuturesPosition
+					(p) => {
```

Comment:
Could we extract this to a separate function as quite a bit of repeated code across these selectors?

Reply:
Good point - I pulled the function into `utils`, logic for updating it is still in the selector.

Reply:
It still duplicates a bit of code, but if we ever need to add some specific logic to update cross margin positions we may need them separate.

Reply:
Not sure I follow why the whole function can't be pulled there as they seem identical?

Reply:
You're right... I was thinking about a future state where we may want to handle these position differently mapping the exact same function might not be ideal. But I don't have a specific reason, so I agree it's best to move the whole thing.